components/microfrontend/pkg/apis/ui/v1alpha1/types.go

```diff
@@ -0,0 +1,72 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type MicroFrontend struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              MicroFrontendSpec `json:"spec"`
+}
+
+// +genclient
+// +genclient:noStatus
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type ClusterMicroFrontend struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              ClusterMicroFrontendSpec `json:"spec"`
+}
+
+func (rem *MicroFrontend) GetObjectKind() schema.ObjectKind {
+	return &MicroFrontend{}
+}
+
+func (rem *ClusterMicroFrontend) GetObjectKind() schema.ObjectKind {
+	return &ClusterMicroFrontend{}
+}
+
+type MicroFrontendSpec struct {
+	Version         string           `json:"version"`
+	Category        string           `json:"category"`
+	ViewBaseUrl     string           `json:"viewBaseUrl"`
+	NavigationNodes []NavigationNode `json:"navigationNodes"`
+}
+
+type ClusterMicroFrontendSpec struct {
+	Placement         string `json:"placement"`
+	MicroFrontendSpec `json:",inline"`
```

Comment:
Please don't use the spec type for MicroFrontend CR directly. Create another `CommonMicroFrontendSpec` and put all common properties there. The `MicroFrontendSpec` will look like:

 ```golang
type MicroFrontendSpec struct {
    CommonMicroFrontendSpec `json:",inline"`
}
```

and the `ClusterMicroFrontendSpec`:

```golang
type ClusterMicroFrontendSpec struct {
	Placement         string `json:"placement"`
	CommonMicroFrontendSpec `json:",inline"
}
```

That's the way the Service Catalog guys do it. Why? It will be easier for extending these two types without making breaking changes. The `ClusterMicroFrontend` CR won't need to contain everything that the `MicroFrontend` CR contains in spec.